### PR TITLE
Eval7 - Update CARLA Object Detection Multimodal Attacks

### PR DESCRIPTION
**Pull request by yusong-tan**
Updated the CARLA object detection multimodal attacks so patch attack is applied to the RGB channels and masked PGD attack is applied to the depth channels. This ensures that the depth perturbation is pixel-accurate wrt to the camera perspective.

### CODE REVIEW

**Comment by jprokos26** [comment]
Causes error when running [eval5 carla_obj_det_multimodal_adversarialpatch_defended.json](https://github.com/twosixlabs/armory/blob/master/scenario_configs/eval5/carla_object_detection/carla_obj_det_multimodal_adversarialpatch_defended.json):

```
Evaluation:   0% 0/1 [00:02<?, ?it/s]
2023-03-21 18:28:25 13s ERROR    armory.scenarios.scenario:evaluate:449 Encountered error during scenario evaluation.
Traceback (most recent call last):
File "/opt/conda/lib/python3.9/runpy.py", line 197, in _run_module_as_main
return _run_code(code, main_globals, None,
│         │     └ {'__name__': '__main__', '__doc__': '\nMain script for running scenarios. Users will run a scenario by calling:\n    armory r...
│         └ <code object <module> at 0x7f9cec9b8df0, file "/workspace/armory/scenarios/main.py", line 1>
└ <function _run_code at 0x7f9ddd5f09d0>
File "/opt/conda/lib/python3.9/runpy.py", line 87, in _run_code
exec(code, run_globals)
│     └ {'__name__': '__main__', '__doc__': '\nMain script for running scenarios. Users will run a scenario by calling:\n    armory r...
└ <code object <module> at 0x7f9cec9b8df0, file "/workspace/armory/scenarios/main.py", line 1>
File "/workspace/armory/scenarios/main.py", line 229, in <module>
run_config(
└ <function run_config at 0x7f9cd197bca0>
File "/workspace/armory/scenarios/main.py", line 145, in run_config
scenario.evaluate()
│        └ <function Scenario.evaluate at 0x7f9cbff684c0>
└ <armory.scenarios.carla_object_detection.CarlaObjectDetectionTask object at 0x7f9ddd4a6730>
> File "/workspace/armory/scenarios/scenario.py", line 440, in evaluate
self._evaluate()
│    └ <function Scenario._evaluate at 0x7f9cbff68430>
└ <armory.scenarios.carla_object_detection.CarlaObjectDetectionTask object at 0x7f9ddd4a6730>
File "/workspace/armory/scenarios/scenario.py", line 430, in _evaluate
self.evaluate_all()
│    └ <function Scenario.evaluate_all at 0x7f9cbff68040>
└ <armory.scenarios.carla_object_detection.CarlaObjectDetectionTask object at 0x7f9ddd4a6730>
File "/workspace/armory/scenarios/scenario.py", line 335, in evaluate_all
self.evaluate_current()
│    └ <function Scenario.evaluate_current at 0x7f9cbff68310>
└ <armory.scenarios.carla_object_detection.CarlaObjectDetectionTask object at 0x7f9ddd4a6730>
File "/workspace/armory/scenarios/scenario.py", line 416, in evaluate_current
self.run_attack()
│    └ <function CarlaObjectDetectionTask.run_attack at 0x7f9cbff68a60>
└ <armory.scenarios.carla_object_detection.CarlaObjectDetectionTask object at 0x7f9ddd4a6730>
File "/workspace/armory/scenarios/carla_object_detection.py", line 52, in run_attack
x_adv = self.attack.generate(
        │    │      └ <function CARLAAdversarialPatchPyTorch.generate at 0x7f9cbed0dee0>
        │    └ <armory.art_experimental.attacks.carla_obj_det_adversarial_patch.CARLAAdversarialPatchPyTorch object at 0x7f9cbebd7c10>
        └ <armory.scenarios.carla_object_detection.CarlaObjectDetectionTask object at 0x7f9ddd4a6730>
File "/workspace/armory/art_experimental/attacks/carla_obj_det_adversarial_patch.py", line 436, in generate
patch,
 _ = super().generate(np.expand_dims(x[i], axis=0), y=[y_gt])
                            │  │           │ │               └ {'labels': array([1, 1, 1, 1, 1, 1, 1, 1, 2, 2]), 'boxes': array([[ 917.,  464.,  955.,  527.],
                            │  │           │ │                        [ 911.,  468.,  940., ...
                            │  │           │ └ 0
                            │  │           └ array([[[[0.39607844, 0.4862745 , 0.5647059 , 1.        , 1.        ,
                            │  │                       1.        ],
                            │  │                      [0.39607844, 0.4862745 ...
                            │  └ <function expand_dims at 0x7f9d553de160>
                            └ <module 'numpy' from '/opt/conda/lib/python3.9/site-packages/numpy/__init__.py'>
File "/opt/conda/lib/python3.9/site-packages/art/attacks/evasion/adversarial_patch/adversarial_patch_pytorch.py", line 606, in generate
_ = self._train_step(images=images, target=target, mask=None)
│    │                  │              └ {'boxes': tensor([[[ 917.,  464.,  955.,  527.],
│    │                  │                         [ 911.,  468.,  940.,  517.],
│    │                  │                         [ 771.,  467.,  788.,  518.]...
│    │                  └ tensor([[[[0.3961, 0.4863, 0.5647, 1.0000, 1.0000, 1.0000],
│    │                              [0.3961, 0.4863, 0.5647, 1.0000, 1.0000, 1.0000],
│    │                         ...
│    └ <function CARLAAdversarialPatchPyTorch._train_step at 0x7f9cbed0dd30>
└ <armory.art_experimental.attacks.carla_obj_det_adversarial_patch.CARLAAdversarialPatchPyTorch object at 0x7f9cbebd7c10>
File "/workspace/armory/art_experimental/attacks/carla_obj_det_adversarial_patch.py", line 120, in _train_step
depth_linear = torch.clamp(
               │     └ <built-in method clamp of type object at 0x7f9da5a16200>
               └ <module 'torch' from '/opt/conda/lib/python3.9/site-packages/torch/__init__.py'>
RuntimeError: The size of tensor a (1280) must match the size of tensor b (3) at non-singleton dimension 3
```

File: armory/art_experimental/attacks/carla_obj_det_adversarial_patch.py [hunk 1]
```diff
@@ -66,51 +85,47 @@ def _train_step(
...
+                    images_depth_linear = rgb_depth_to_linear(
+                        images_depth[:, 0, :, :],
+                        images_depth[:, 1, :, :],
+                        images_depth[:, 2, :, :],
+                    )
+                    depth_linear = torch.clamp(
+                        images_depth_linear + depth_linear,
+                        min=self.min_depth,
+                        max=self.max_depth,
+                    )
```
Comment (jprokos26) [inline armory/art_experimental/attacks/carla_obj_det_adversarial_patch.py:124]:
```suggestion
                    if depth_linear.shape != self.min_depth.shape:
                        assert torch.diff(self.min_depth, dim=-1).sum() == 0
                        assert torch.diff(self.max_depth, dim=-1).sum() == 0
                        depth_linear = torch.clamp(
                            images_depth_linear + depth_linear,
                            min=self.min_depth[..., 0],
                            max=self.max_depth[..., 0],
                        )
                    else:
                        depth_linear = torch.clamp(
                            images_depth_linear + depth_linear,
                            min=self.min_depth,
                            max=self.max_depth,
                        )
```

Comment (yusong-tan) [reply armory/art_experimental/attacks/carla_obj_det_adversarial_patch.py]: @jprokos26 my latest commit fixes the issue.

**Comment by jprokos26** [comment]
Attack confirmed to be working with [eval5 carla_obj_det_multimodal_adversarialpatch_defended.json](https://github.com/twosixlabs/armory/blob/master/scenario_configs/eval5/carla_object_detection/carla_obj_det_multimodal_adversarialpatch_defended.json) and [eval6 carla_obj_det_multimodal_adversarialpatch_defended.json](https://github.com/twosixlabs/armory/blob/master/scenario_configs/eval6/carla_overhead_object_detection/carla_obj_det_multimodal_adversarialpatch_defended.json).

**Review by jprokos26 — APPROVED** [review]
Attack confirmed to be working with [eval5 carla_obj_det_multimodal_adversarialpatch_defended.json](https://github.com/twosixlabs/armory/blob/master/scenario_configs/eval5/carla_object_detection/carla_obj_det_multimodal_adversarialpatch_defended.json) and [eval6 carla_obj_det_multimodal_adversarialpatch_defended.json](https://github.com/twosixlabs/armory/blob/master/scenario_configs/eval6/carla_overhead_object_detection/carla_obj_det_multimodal_adversarialpatch_defended.json).

**Pull request merged**